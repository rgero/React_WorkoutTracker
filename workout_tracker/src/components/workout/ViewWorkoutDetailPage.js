import React, { useContext, useEffect, useState } from 'react';
import {useParams} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';

import DateFormatter from '../../helpers/DateFormatter';

const ViewWorkoutDetailsPage = ()=> 
{
    const {state, fetchWorkoutByID} = useContext(WorkoutContext);
    const [isLoaded, setLoaded] = useState(false);
    const { id } = useParams();

    useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            await fetchWorkoutByID(id);
            setLoaded(true);
        }
        GetWorkoutByID(id);
    }, []);

    if (state.workoutDate)
    {
        return (
            <Container fluid="md">
                <h1>{DateFormatter(state.workoutDate)}</h1>
            </Container>
            )
    } else {
        return (<div>Loading</div>)
    }
}

export default ViewWorkoutDetailsPage;