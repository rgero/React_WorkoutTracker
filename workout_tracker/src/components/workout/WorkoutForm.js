import React, {useState, useEffect} from 'react';

import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';

import '../../styles/components/inputs.css'
import '../../styles/components/WorkoutForm.css';

import { ExerciseList } from '../exercise/ExerciseList';
import { ExerciseForm } from '../exercise/ExerciseForm';
import DateFormatter from '../../helpers/DateFormatter';

export const WorkoutForm = ({workout = {}, errorMessage, onSubmit})=> 
{
    const [workoutDate, setWorkoutDate] = useState(workout.workoutDate ? DateFormatter(new Date(workout.workoutDate)) : DateFormatter(new Date()));
    const [exerciseList, setExerciseList] = useState(workout.exerciseList ? workout.exerciseList : []);
    const [notes, setNotes] = useState(workout.notes ? workout.notes : "");
    const [error, setError] = useState("");

    const addExercise = (exercise) => {
        setExerciseList([...exerciseList, exercise]);
    }
    
    const deleteExercise = (index) => {
        let newExerciseList = exerciseList;
        newExerciseList.splice(index, 1);
        setExerciseList( [...newExerciseList] );
    }

    const submitWorkout = (e) => {
        e.preventDefault();
        if (!workoutDate)
        {
            setError("Need a workout date");
        }
        if (exerciseList.length < 1)
        {
            setError("Need at least one exercise");
        }
        if (error)
        {
            return;
        }
        let workout = {
            workoutDate,
            exerciseList,
            notes
        }

        onSubmit(workout);
    }

    return (
        <Container fluid="md">
            <Form id="addWorkout" onSubmit={submitWorkout}>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Date</InputGroup.Text>
                    <Form.Control type="date" value={workoutDate} onChange={e => setWorkoutDate(e.target.value)}/>
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Notes</InputGroup.Text>
                    <Form.Control as="textarea" value={notes} rows={3} onChange={e => setNotes(e.target.value)} />
                </InputGroup>
            </Form>
            <div className="exerciseForm">
                <ExerciseForm onSubmit={addExercise}/>
            </div>
            <div className="exerciseList">
                <ExerciseList exerciseList={exerciseList} onDelete={deleteExercise}/>
            </div>
            <Button form="addWorkout" variant="outline-secondary" type="submit">
                    Submit Workout
            </Button>
        </Container>

    )
}