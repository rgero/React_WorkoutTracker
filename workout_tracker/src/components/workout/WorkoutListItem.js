import React from 'react';
import Card from 'react-bootstrap/Card';

import DateFormatter from '../../helpers/DateFormatter';

export const WorkoutListItem = ({workout}) => {
    let {workoutDate, notes, exerciseList} = workout;
    workoutDate = new Date(workoutDate);

    return (
        <div>
            <Card bordered="true">
                <Card.Body>
                    <Card.Title>{DateFormatter(workoutDate)}</Card.Title>
                    <Card.Text>{notes}</Card.Text>
                    <Card.Text>Total Exercises: {exerciseList.length}</Card.Text>
                </Card.Body>
            </Card>
        </div>
    )
}