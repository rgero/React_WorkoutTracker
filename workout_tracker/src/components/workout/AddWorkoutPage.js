import React, { useContext, } from 'react';
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from './WorkoutForm';

export const AddWorkoutPage = ()=> 
{
    const {state, createWorkout} = useContext(WorkoutContext);
    return (
        <Container fluid="md">
            <h1>Create a new workout</h1>
            <WorkoutForm
                errorMessage={state.errorMessage}
                onSubmit={({workout})=> createWorkout({workout})}
            />
        </Container>
    )
}