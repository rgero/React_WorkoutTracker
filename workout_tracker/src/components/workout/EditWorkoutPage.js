import React, { useContext, useEffect, useState } from 'react';
import {useParams} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from './WorkoutForm';

export const EditWorkoutPage = ( )=> 
{
    const {state, fetchWorkoutByID, updateWorkout} = useContext(WorkoutContext);
    const [isLoaded, setLoaded] = useState(false);
    const { id } = useParams();

    useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            await fetchWorkoutByID(id);
            setLoaded(true);
        }
        GetWorkoutByID(id);
    }, []);

    if (state.workoutDate)
    {
        return (
            <Container fluid="md">
                <h1>Edit a workout</h1>
                <WorkoutForm
                    workout={state}
                    errorMessage={state.errorMessage}
                    onSubmit={(workout)=> updateWorkout(workout)}
                />
            </Container>
            )
    } else {
        return (<div>Loading</div>)
    }
}