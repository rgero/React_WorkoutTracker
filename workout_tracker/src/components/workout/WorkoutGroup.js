import React from 'react';
import Accordion from 'react-bootstrap/Accordion';
import { WorkoutListItem } from "./WorkoutListItem";
import { GetMonthName, GetOrderedValues } from '../../helpers/WorkoutListProcesser';

const WorkoutGroup = ({workoutList, descendingOrder}) => {
    let months = GetOrderedValues(workoutList, descendingOrder);
    return (
        <div>
            {
                months.map((month, index)=> (
                    <Accordion defaultActiveKey={index} key={`WorkoutGroup_${index}`}>   
                        <Accordion.Item>
                            <Accordion.Header>{GetMonthName(month)}</Accordion.Header>
                            <Accordion.Body>
                                {
                                    workoutList[month].map((workout, index) => (
                                            <WorkoutListItem workout={workout} key={`WorkoutListItem_${index}`} />
                                        ))
                                }
                            </Accordion.Body>
                        </Accordion.Item>
                    </Accordion>
                ))
            }
        </div>
      );
}

export default WorkoutGroup;