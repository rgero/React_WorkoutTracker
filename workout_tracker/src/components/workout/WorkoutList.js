import React from "react";
import Container from "react-bootstrap/Container";

import Collapsible from "react-collapsible";
import {GetOrderedValues, ProcessWorkoutList} from "../../helpers/WorkoutListProcesser";
import WorkoutGroup from "./WorkoutGroup";


export const WorkoutList = ({workoutList=[]}) => 
{

    let organizedList = ProcessWorkoutList(workoutList);
    let yearOrder = GetOrderedValues(organizedList, true);
    return (
        <Container>
            { 
                Object.keys(organizedList).length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Workouts</span>
                    </div>
                ) : (
                    <div>
                        
                        {
                            yearOrder.map((year, index)=> (
                                <Collapsible trigger={year}>
                                    <WorkoutGroup {...organizedList[year]} />
                                </Collapsible>
                            ))
                        }
                    </div>
                )
            }
        </Container>
    )
}