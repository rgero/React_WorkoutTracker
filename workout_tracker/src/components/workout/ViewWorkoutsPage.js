import React, {useContext, useEffect} from 'react';
import Container from 'react-bootstrap/Container';
import { WorkoutList } from './WorkoutList';
import {Context as WorkoutContext} from '../../context/WorkoutContext';

export const ViewWorkoutsPage = ()=> 
{
    const {state, fetchWorkouts} = useContext(WorkoutContext);

    useEffect(()=> {
        const processWorkouts = async () => {
            await fetchWorkouts();
        }
       processWorkouts();
    }, []);

    return (
        <Container>
            <div>Workouts</div>
            <WorkoutList workoutList={state} />
        </Container>
    )
}