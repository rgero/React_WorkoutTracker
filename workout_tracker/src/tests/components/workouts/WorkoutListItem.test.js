import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow'
import {WorkoutListItem} from '../../../components/workout/WorkoutListItem';

import { TestWorkouts } from '../../testData';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn()
}));

let realUseContext;
let useContextMock;

beforeEach(()=> {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();

    useContextMock.mockReturnValue({
        deleteWorkout: jest.fn()
    });
})

afterEach(()=> {
    React.useContext = realUseContext;
})

describe("Workout List Item - Render tests", ()=> {
    test("Has Data, matches snapshot", ()=> {
        const pageRender = new ShallowRenderer().render(
            <WorkoutListItem workout={TestWorkouts[0]} index={1} />
        );
        expect(pageRender).toMatchSnapshot();
    })
})