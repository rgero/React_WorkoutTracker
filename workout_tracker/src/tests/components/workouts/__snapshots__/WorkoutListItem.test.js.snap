// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout List Item - Render tests Has Data, matches snapshot 1`] = `
<Accordion
  defaultActiveKey={1}
>
  <AccordionItem>
    <AccordionHeader>
      2023-03-08
    </AccordionHeader>
    <AccordionBody>
      <Container
        fluid={false}
      >
        <Row>
          <Col
            sm={4}
          >
            Notes
          </Col>
          <Col>
            Testing Notes for Testing
          </Col>
        </Row>
        <Row
          className="pb-4"
        >
          <Col
            sm={4}
          >
            Total Exercises
          </Col>
          <Col>
            2
          </Col>
        </Row>
        <Row>
          <Col
            sm={4}
          >
            <Button
              active={false}
              aria-label="viewWorkout"
              disabled={false}
              onClick={[Function]}
              variant="outline-primary"
            >
              View
            </Button>
          </Col>
          <Col
            sm={4}
          >
            <Button
              active={false}
              aria-label="editWorkout"
              disabled={false}
              onClick={[Function]}
              variant="outline-primary"
            >
              Edit
            </Button>
          </Col>
          <Col
            sm={4}
          >
            <Button
              active={false}
              aria-label="deleteWorkout"
              disabled={false}
              onClick={[Function]}
              variant="outline-danger"
            >
              Delete
            </Button>
          </Col>
        </Row>
      </Container>
    </AccordionBody>
  </AccordionItem>
</Accordion>
`;
